x/wasm/keeper/query_plugins.go

```diff
@@ -56,18 +56,26 @@ func (q QueryHandler) Query(request wasmvmtypes.QueryRequest, gasLimit uint64) (
 	}()
 
 	res, err := q.Plugins.HandleQuery(subCtx, q.Caller, request)
-	// Error mapping
+	if err == nil {
+		// short-circuit, the rest is dealing with handling existing errors
+		return res, nil
+	}
+
+	// special mappings to system error
 	var noSuchContract *types.ErrNoSuchContract
 	if ok := errors.As(err, &noSuchContract); ok {
-		return res, wasmvmtypes.NoSuchContract{Addr: noSuchContract.Addr}
+		err = wasmvmtypes.NoSuchContract{Addr: noSuchContract.Addr}
 	}
-	// Otherwise redact all (we can theoretically redact less in the future)
-	if err != nil {
+
+	// Redact all that are not system errors
+	// SystemErrors must be created in x/wasm and we can ensure determinism
+	// (we can theoretically redact less in the future, but this is a first step to safety)
+	if wasmvmtypes.ToSystemError(err) == nil {
```

Comment:
Would it make sense to exclude system errors in https://github.com/CosmWasm/wasmd/blob/master/x/wasm/keeper/msg_dispatcher.go#L139 as well?

Reply:
Good point.

I can actually generalise the redactError logic